scripts/proto/lbann_proto.py

```diff
@@ -349,68 +475,73 @@ def __init__(self, **kwargs):
                 setattr(self, field_name, kwargs[field_name])
             else:
                 setattr(self, field_name, None)
-    # Method for adding to the protobuf model.
-    def proto(self, model):
-        callback = self._add_callback(model)
-        callback_field = getattr(callback, message_name)
-        callback_field.SetInParent()  # Create empty message.
+
+    # Method for exporting a protobuf message.
+    def export_proto(self):
+        proto = Callback.export_proto(self)
+        callback_message = getattr(proto, callback_field_name)
+        callback_message.SetInParent() # Create empty message
         for field_name in field_names:
             v = getattr(self, field_name)
             if v is not None:
-                # Handle repeated fields.
-                if type(v) is list:
-                    field = getattr(callback_field, field_name)
-                    field.extend(v)
-                else:
-                    setattr(callback_field, field_name, v)
-    _proto_callbacks[type_name] = type(type_name, (Callback,),
-                                       {'__init__': __init__, 'proto': proto})
-
-# Generate Callback sub-classes.
+                if type(v) is list: # Repeated field
+                    getattr(callback_message, field_name).extend(v)
+                else: # Singular field
+                    setattr(callback_message, field_name, v)
+        return proto
+
+    # Create sub-class.
+    return type(type_name, (Callback,),
+                {'__init__': __init__, 'export_proto': export_proto})
+
+# Generate Callback sub-classes from lbann.proto
+# Note: The list of skip fields must be updated if any new fields are
+# added to the Callback message in lbann.proto
+_skip_fields = set([])
+_generated_classes = {}
 for field in lbann_pb2.Callback.DESCRIPTOR.fields:
-    _create_proto_callback(field.message_type.name, field.name)
-_add_to_module_namespace(_proto_callbacks)
+    if field.name not in _skip_fields:
+        type_name = field.message_type.name
+        _generated_classes[type_name] = _create_callback_subclass(type_name)
+_add_to_module_namespace(_generated_classes)
 
-def traverse_model(first_module, func):
-    """Traverse a model and apply func to each layer.
+# ==============================================
+# Export models
+# ==============================================
 
-    first_module is the first module in the model (i.e. the input layer).
-    The model is traversed in a depth-first order except that no layer is
-    visited until all its ancestors have been visited.
+def save_model(filename, mini_batch_size, epochs,
+               layers=[], weights=[], objective_function=None,
```

Comment:
Does it ever make sense to pass no layers? If not, maybe we shouldn't default this to `[]`.

Reply:
I could imagine someone doing something with just weights and objective functions, but that would be a very weird use-case.